### ISSUE TITLE
ES6 Modules

### ISSUE THREAD

**Issue by IlyaKhD** [DevExpress/DevExtreme#14351]
## The Problem

Tree Shaking optimizes bundles that use ES6 modules (see descriptions at [webpack.js.org](https://webpack.js.org/guides/tree-shaking/) and [rollupjs.org](https://rollupjs.org/guide/en/#tree-shaking)). DevExtreme is available as a set of CommonJS modules and cannot be optimized in this way.
 
We also have multiple reported issues related to CommonJS modules, Tree Shaking, and large bundle size:
 
- https://github.com/DevExpress/devextreme-react/issues/84
- https://github.com/DevExpress/devextreme-angular/issues/1086
- https://github.com/DevExpress/devextreme-angular/issues/811

In addition, Angular recommends [using ES6 modules in your entire application](https://angular.io/guide/build#configuring-commonjs-dependencies).

## The Proposed Solution

To deliver a set of ES6 modules in addition to CommonJS modules. This step solves all the described problems while providing backward compatibility. It also makes future enhancements like [Improved Code Splitting](https://js.devexpress.com/Roadmap/#ArchitecturalEnhancements) more efficient.

No changes in your projects are required. The new ES6 modules will be used instead of the old CommonJS modules automatically.

### Take a Quick Poll

[Do you find Tree Shaking useful?](https://forms.gle/bnBRMUqfMmzx78Eu8)

[Have you tried using multiple bundles in any of your projects (e.g. a bundle per page)?](https://forms.gle/kNXBZo8cHnRPciPY8)

## Get Update Notifications

Subscribe to this thread, or to our [Facebook](https://www.facebook.com/DevExpress.DevExtreme/) and [Twitter](https://twitter.com/devextreme) accounts, for updates on this topic.

**Comment by hakimio** [comment]
Can we already test this or will it be in the next alpha release?

**Comment by pedrofurtado** [comment]
This amazing feature will be available in next 20.2 official release? 🎉

**Comment by IlyaKhD** [comment]
That's right.
I think it will be available for testing a bit earlier - let's say, in 20.2 Beta. If so, I will post an update in this thread.

**Comment by erdoganmegma** [comment]
That's awesome! Any Eta updates?

**Comment by softboy99** [comment]
Sounds great.

**Comment by BoBiene** [comment]
@IlyaKhD When do you think we can try this out in an alpha-build?

**Comment by MaximKudriavtsev** [comment]
Hi,

Our work on this feature is at the final stage. We will update this topic once we publish an NPM package with it.

**Comment by BoBiene** [comment]
@MaximKudriavtsev is it part of the 20.2.2-beta?

**Comment by MaximKudriavtsev** [comment]
Hi @BoBiene,

Unfortunately no, the Beta version doesn't include ES6 modules. We have some unexpected implementation issues, and now we are fixing them.

We would like to publish one more Beta version with this feature before the official release. I will notify you once we do so.

**Comment by softboy99** [comment]
Does it means if this are finished, tools like Vite Scafford for Vue3 also be supported? Vite also depends on ES6 modules.

**Comment by IlyaKhD** [comment]
Unfortunately, Vite requires that all packages have ES6 modules. DevExtreme depends on [jszip](https://www.npmjs.com/package/jszip) which provides CommonJS modules only.

**Comment by softboy99** [comment]
Hi
https://github.com/Stuk/jszip/issues/349, it supports es6 import

**Comment by hakimio** [comment]
@softboy99 the problem with jszip is not with the way you import it but how the library itself is structured. It's a CommonJS module.

**Comment by softboy99** [comment]
Hi,
https://www.npmjs.com/package/@progress/jszip-esm. It is forked by telerik

**Comment by IlyaKhD** [comment]
Hi everyone,

We'd like to make an update on ES6 modules support.

The good news is that we've finished all the major tasks related to this feature. However, when we tested our new module set with some of the existing DevExtreme projects, we faced unexpected issues that might break our customers' projects. So, we consider it to be too risky to make it part of our stable v20.2 release.

Since it's very important to let you test these new modules in your projects, we plan to include them in the very first v21.1 alpha release that we expect to publish in a few weeks.

The next step in our bundle size optimization journey is to organize DevExtreme into smaller modules. For instance, to split our DataGrid into multiple parts based on different use cases. It will enhance the overall tree shaking and code splitting. We would also appreciate it if you share your cases when unused/unexpected code is bundled while it shouldn't. It'll help us improve the most important parts first.

Please stay tuned and thank you for your collaboration.

**Comment by javzwin** [comment]
So the only solution so far to avoid these warnings is "allowedCommonJsDependencies"??

**Comment by hakimio** [comment]
@javzwin until ES6 modules are released, yes, it is.

**Comment by mrparag94** [comment]
Does this 21.1 Branch supports es6 import..?

**Comment by IlyaKhD** [comment]
Hi, @mrparag94, sorry for the late response. 

I'm not sure you mean that, but you can use ES6 `import` statements with all actual DevExtreme versions, e.g., v20.2.
About the 21_1 branch, we're reorganizing the package structure, which is important for bundle optimization.

**Comment by Rai-Rai** [comment]
When is the release of 21.1 planned?

**Comment by hakimio** [comment]
@Rai-Rai the major DX releases usually come in 6 months intervals, so I would guess May 2021.

**Comment by Semigradsky** [comment]
Do you have plans for moving `devexpress-diagram`, `devexpress-gantt`, `devextreme-quill` to `optionalDependencies`? Or just remove from `dependencies`

**Comment by softboy99** [comment]
Hi,
Does the 21.1.1-sprint3 is your 21.1 alpha?

**Comment by 101arrowz** [comment]
> Unfortunately, Vite requires that all packages have ES6 modules. DevExtreme depends on [jszip](https://www.npmjs.com/package/jszip) which provides CommonJS modules only.

An alternative to `jszip-esm` would be [`fflate`](https://github.com/101arrowz/fflate). In addition to supporting ESM, it is much faster, much smaller, and has [these added benefits](https://github.com/davidramos-om/zipcelx-on-steroids/pull/1#issue-545029437).

**Comment by artem-kurchenko** [comment]
Hi,

@softboy99

Though the current source code (21_1) uses ES 6 modules, there is still a number of issues that we need to address before telling you that everything is ready for the first preview. We're also still testing dependent packages and common use cases. Once we are done with these tasks, we'll present all changes and release a package for public testing. Please watch for our updates here, and you'll get a notification about the early preview.

@Semigradsky

Thank you for your suggestion. We're considering one of these options.

**Comment by ducmerida** [comment]
@101arrowz @IlyaKhD  Vite converts CommonJS dependencies to ESM in a Pre-Building step
![imagen](https://user-images.githubusercontent.com/14205169/104737063-1401ff80-5744-11eb-86e4-b6eb8ede3ef4.png)
https://vitejs.dev/guide/dep-pre-bundling.html

**Comment by MaximKudriavtsev** [comment]
Hi everyone,

We're happy to announce that ES6 modules are available in DevExtreme NPM distribution (starting [21.1 Sprint 8](https://github.com/DevExpress/DevExtreme/releases/tag/21.1.1-sprint.8)).

We've done some internal testing and can see better tree shaking results. We built bundles with certain individual components to compare the final bundle sizes. Here are our results: 

| Component | cjs size | esm size |
| ------- | -------- | ------- |
| Button | 249Kb | 191Kb |
| Form | 994Kb | 845Kb |
| Scheduler | 2Mb | 1,7Mb |
| DataGrid | 1,9Mb | 1,6Mb |

The next step in this journey is better code-splitting. Smaller independent modules will allow you to produce smaller code chunks that contain only the code you use.

Please try our ES6 modules and let us know if you face any issues. Any further feedback on better code-splitting is appreciated.

### NPM Packages

```
devextreme@21.1.1-alpha-21064-0310

devextreme-angular@21.1.1-alpha-21064-0310
devextreme-react@21.1.1-alpha-21064-0310
devextreme-vue@21.1.1-alpha-21064-0310
```

**Comment by AlissonRS** [comment]
@MaximKudriavtsev thanks for the update. I'll talk with my team about upgrading to 21.1 as soon as possible, and will let you know if we face any issues.

The 4 components you mentioned are just for benchmarking right? I'm assuming this improvement also applies to PivotGrid and other components.

**Comment by MaximKudriavtsev** [comment]
Hi @AlissonRS,

> The 4 components you mentioned are just for benchmarking right?

You are right, these components are examples. We ship ES6 modules for all our components.

**Comment by bobskigit** [comment]
Can we publish the localization messages as es6 modules?

Warning: culture.service.ts depends on 'devextreme/dist/js/localization/dx.messages.fr'. CommonJS or AMD dependencies can cause optimization bailouts.
For more info see: https://angular.io/guide/build#configuring-commonjs-dependencies

**Comment by MaximKudriavtsev** [comment]
Hi @bobskigit,

Thank you for your feedback. We will publish localization messages as es6 modules in future releases.

Could you please share your project bundle size with and without the ES6 modules feature? Also, any additional feedback would be appreciated. This information helps us plan future enhancements.

**Comment by beeing** [comment]
Just wondering if there's any improvements planned for the CSS part as well? Thanks.

**Comment by hakimio** [comment]
@beeing https://github.com/DevExpress/DevExtreme/issues/13654

**Comment by bobskigit** [comment]
My bundle size is reduced by 308Kb in a 2.8Mb bundle when using es6 modules. Nice work.

**Comment by MaximKudriavtsev** [comment]
Hi @bobskigit,

> My bundle size is reduced by 308Kb in a 2.8Mb bundle when using es6 modules. Nice work.

Thank you for sharing the size of your project bundles. We appreciate it.

> Can we publish the localization messages as es6 modules?

We researched the issue and found a solution. Import `.json` localization files and load them by calling the [loadMessages](https://js.devexpress.com/Documentation/21_1/ApiReference/Common/Utils/localization/#loadMessagesmessages) function. You can refer to the `NPM` section in the [Dictionaries](https://js.devexpress.com/Documentation/21_1/Guide/Common/Localization/#Dictionaries) article for more details. Please let us know if this resolves the issue on your side.

**Comment by softboy99** [comment]
Hi
for 21.1.2-beta, cannot be used with vite. The following is the error message:
Uncaught SyntaxError: The requested module '/node_modules/.vite/devextreme_ui_dialog.js?v=8e88dbc9' does not provide an export named 'default'

**Comment by alexserov** [comment]
Hello @softboy99,
It seems that you're trying to use the default export instead of the named one in your app. Try the syntax from our [documentation](https://js.devexpress.com/Documentation/ApiReference/Common/Utils/ui/dialog/#customoptions) and share you results with me.

**Comment by salemu01** [comment]
Hello,
 I've been waiting impatiently for this and tested the 21.1.2-beta.
I hoped i would have noticed much bigger difference, but it's still at least a first noticeable difference :
  




<meta name="ProgId" content="Excel.Sheet">
<meta name="Generator" content="Microsoft Excel 15">
<link id="Main-File" rel="Main-File" href="file:///C:/Users/slu/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel="File-List" href="file:///C:/Users/slu/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

main.js with no   devextreme | 2,4mb
-- | --
main.js with devextreme 20.x | 3,8mb
main.js with devextreme 21.1.2 | 3,6mb

As the work for smaller independent modules continues, i hope we get to see a much more reduced size on the final project bundle.

(Haven't noticed any particular issues on this beta version, except a single CustomStore issue that i created and already solved by the team. Thank you.)

**Comment by dxbykov** [comment]
@salemu01 Thanks for sharing the numbers. Note that Tree Shaking can't magically 'compress' the code. It just eliminates the code that is not used in your app. So, if you use everything, there won't be any difference in the final bundle size.

We would like to examine your particular case. Could you please specify what components/features you are using (or not using) in your app? If you could specify what's included in the final bundle that shouldn't, it would be even more helpful.

**Comment by salemu01** [comment]
@dxbykov Thanks you. I will join here the output of bundle analyser, as well as all my imports, if you could point out any possible improvement.

**Comment by salemu01** [comment]
@dxbykov After first analysis, everything seems rather normal to me. On my lazy loaded module (which isn't the concern) i see only dx modules that i imported in my project. 
On the main.js bundle, although i hoped i may end up with a lighter bundle, i suppose i have the minimum necessary imports, and if some weren't, i don't think it would be of consequent size. 

Here is as an overview, my lazy loaded imported modules:
![image](https://user-images.githubusercontent.com/84332174/118985497-c78f3080-b97e-11eb-9dea-8a08a735ab84.png)

And here is in the main.js bundle:

![image](https://user-images.githubusercontent.com/84332174/118985946-32406c00-b97f-11eb-8807-9d7b1ccc4880.png)
![image](https://user-images.githubusercontent.com/84332174/118987926-0625ea80-b981-11eb-8654-774ae870cb6e.png)

Thank you.

**Comment by softboy99** [comment]
Hi,
after upgrade to 21.1.3, cannot start with the following error,which are ok with 20.2.x:
component.js:65 Uncaught (in promise) TypeError: callbacks_1.default is not a function
    at Proxy.data (component.js:65)
    at resolveData (runtime-core.esm-bundler.js:6046)
    at runtime-core.esm-bundler.js:5860
    at Array.forEach (<anonymous>)
    at applyOptions (runtime-core.esm-bundler.js:5860)
    at finishComponentSetup (runtime-core.esm-bundler.js:6636)
    at setupStatefulComponent (runtime-core.esm-bundler.js:6567)
    at setupComponent (runtime-core.esm-bundler.js:6503)
    at mountComponent (runtime-core.esm-bundler.js:4206)
    at processComponent (runtime-core.esm-bundler.js:4182)

vite 2.3.4 +  vue3.0.11 + devexreme 21.1.3

**Comment by alexserov** [comment]
Hi @softboy99, 
It seems to be a Vite 2.3.x issue, you can track the https://github.com/vitejs/vite/issues/3399 for more info and updates.

As a workaround, I suggest you downgrade Vite to 2.2.4

**Comment by softboy99** [comment]
Hi,
afer upgrade to 21.1.3 with vite 2.3.8. It's ok now. Thanks. But some Icons cannot be displayed dx-icon-menu,dx-icon-folder,dx-icon-dropdown-editor etc.  When use vue-cli, its ok, is it vite issue or devextreme issue?

**Comment by babich-a** [comment]
Hi @softboy99 . Please check the following breaking change [https://supportcenter.devexpress.com/ticket/details/t977747/theme-files-are-shipped-with-the-utf-8-encoding-starting-with-v20-2](https://supportcenter.devexpress.com/ticket/details/t977747/theme-files-are-shipped-with-the-utf-8-encoding-starting-with-v20-2). If it is not related to the issue, we need more info on how dx icons are rendered.

**Comment by softboy99** [comment]
Hi，
when donwgrade to 20.2.7, it is ok.

**Comment by babich-a** [comment]
There have been no serious style changes between 20.2.7 and 21.1.3. Can you send us a simple project or repository that reproduces the issue?

**Comment by softboy99** [comment]
Hi,
i need time to prepare an sample project for you. But i found in the chrome devtools,->sources->node_modules only dx.common.css are there. I'm not sure if it is vite issue or not.

**Comment by babich-a** [comment]
@softboy99  BTW dx.common.css has no styles anymore in 21.1 and the 20.2 last versions.

> i need time to prepare an sample project 

It will be great if you create a separate issue with the 'styles' problem.

**Comment by softboy99** [comment]
Hi,
you can see the demo project:   [https://github.com/softboy99/vite-devexteme21_1.git](url)

**Comment by babich-a** [comment]
@softboy99 thanks.
I reproduce this issue both for v20.2.7 and v21.1.3.
It looks like some vite css plugin has an issue with a comma-separated list.
We have the following value in `theme.base.css` and `theme.additional.css` (21.1.3):

```
src: local("DevExtreme Material Icons"),local("devextreme_material_icons"),url("../../../node_modules/devextreme/dist/css/icons/dxiconsmaterial.woff2") format("woff2"),url("../../../node_modules/devextreme/dist/css/icons/dxiconsmaterial.woff") format("woff"),url("../../../node_modules/devextreme/dist/css/icons/dxiconsmaterial.ttf") format("truetype");
```

v20.2.7 has no quotes in url().

After I run `npm run build` or `npm run dev`, I see the following rule in the stylesheet (`dist\assets\index.<hash>.css`):
```
src:local("DevExtreme Material Icons"),local("devextreme_material_icons") url(/assets/dxiconsmaterial.be9f68c6.woff2) format("woff2") url(/assets/dxiconsmaterial.879a8d81.woff) format("woff") url(/assets/dxiconsmaterial.5a8e9d31.ttf) format("truetype");
```
It has no commas between items though there should be commas according to the [css specification](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/src).

**Comment by softboy99** [comment]
Hi,
Thanks. The right version is 20.2.6. We can post an issule to vite community.

**Comment by AlisherAmonulloev** [comment]
Thank you for your reply. Please do post the issue to the vite community. Thank you in advance!

**Comment by iknowmagic** [comment]
Vite is giving this error: /node_modules/devextreme-vue/popup.js?v=f3801dc8' does not provide an export named 'DxPopup'

With devextreme 21.1.5 + devextreme-vue 21.1.5 and vite 2.4.4.

I guess that DevExtreme is not supporting vitejs at the moment?

Thanks.

**Comment by jacksonrr3** [comment]
@iKnowMagic 
Hi,
I couldn't reproduce the error with the following configuration
- devextreme@21.1.5
- devextreme-vue@21.1.5
- vite@2.4.4 
- vue@3.2.11

Would you please attach an  example with popup that demonstrates the issue?

**Comment by iknowmagic** [comment]
Many thanks, @jacksonrr3

I created a new Vue / ViteJS app, added DevExtreme and I did not see any errors while using a DevExtreme component.

Unfortunately, at the moment I am unable to provide a full environment that exemplifies the issue.

 ![image](https://user-images.githubusercontent.com/8712362/133954417-c6d61d71-b7fd-4018-b853-ce6b6c872875.png)

**Comment by jacksonrr3** [comment]
@iKnowMagic 
Thank you for the update. Unfortunately we cannot tell what goes wrong without an example. Right now I can only recommend you double-check versions of Vue, Vite and devextreme and update them to the configuration I mentioned above.

**Comment by ksercs** [comment]
Thanks you everybody, who gave us feedback on this feature.  I'm closing this thread. In the case of bugs or questions, feel free to create a new ticket in our [Support Center](https://www.devexpress.com/Support/Center/).
